{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "172ab128",
   "metadata": {},
   "source": [
    "# Explore Sequence Diversity\n",
    "\n",
    "This notebook is for exploring protein sequence diversity in a set of CAZymes, e.g. a CAZy family.\n",
    "\n",
    "Use `cazy_webscraper` to retrieve the protein sequences for your dataset of interest, and extract the protein sequences to a FASTA file. Use the provided `run_blastp.sh` or `run_diamond.sh` bash scripts to use BLASTP+ or DIAMOND, respectively, to perform an all-vs-all sequence comparison across your dataset of interest. \n",
    "\n",
    "This notebook takes as input the output from BLASTP+/DIAMOND and visualises the data. \n",
    "\n",
    "Feel free to use this notebook as a template to perform further analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98147226",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4f178c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import time\n",
    "\n",
    "import mechanicalsoup\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "from requests.exceptions import ConnectionError, MissingSchema\n",
    "from tqdm.notebook import tqdm\n",
    "from urllib3.exceptions import HTTPError, RequestError"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3f1a8ac",
   "metadata": {},
   "source": [
    "## Constants\n",
    "\n",
    "Define proteins of interest, e.g. proteins to be explored in the lab. These will be highlighed on the resulting clustermaps and heatmaps.\n",
    "\n",
    "The `dict` uses the group name (e.g. a CAZy family) as the key, and is valued by a list of the NCBI protein version accessions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4f7bf5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "CANDIDATES = {\n",
    "    'grp_name': ['protein_acc']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36a2115b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the colour palettes for annotating proteins\n",
    "PALETTE = sns.color_palette(['#425df5', '#eb8913', '#19bfb4', '#db0d4e', '#15ab62', '#ffffff'])\n",
    "PALETTE_DICT = {\n",
    "    'cand': '#425df5',  # candidates\n",
    "    'struct': '#eb8913',  # protein with structures in RCSB PDB\n",
    "    'structCand': '#19bfb4',  # candidates with structures in RCSB PDB\n",
    "    'func': '#db0d4e',  # candidates listed as 'characterised' in CAZy\n",
    "    'funcCand': '#15ab62',  # proteins listed as 'characterised' in CAZy\n",
    "    'nothing': '#ffffff',  # nothing to note about this protein\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cecf510",
   "metadata": {},
   "source": [
    "## Identify proteins of interest\n",
    "\n",
    "The clustermaps and heatmaps generated to visualise the sequence diversity across the CAZyme database, can be annotated to identify proteins of interest (e.g. candidates, see above), functionally characterised proteins (proteins listed as 'characterised' in CAZy) and structurally characterised proteins (listed under 'structures' in CAZy and also be retrieved from UniProt using `cazy_webscraper`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "888ecfed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this code block before retrieving data from CAZy\n",
    "\n",
    "def browser_decorator(func):\n",
    "    \"\"\"Decorator to re-invoke the wrapped function up to 10 times.\"\"\"\n",
    "\n",
    "    def wrapper(*args, **kwargs):\n",
    "        tries, success, err = 0, False, None\n",
    "\n",
    "        while not success and (tries < kwargs['max_tries']):\n",
    "            try:\n",
    "                response = func(*args, **kwargs)\n",
    "            except (\n",
    "                ConnectionError,\n",
    "                HTTPError,\n",
    "                OSError,\n",
    "                MissingSchema,\n",
    "                RequestError,\n",
    "            ) as err_message:\n",
    "                if (tries < kwargs['max_tries']):\n",
    "                    print(\n",
    "                        f\"Failed to connect to CAZy on try {tries}/{kwargs['max_tries']}.\\n\"\n",
    "                        f\"Error: {err_message}\"\n",
    "                        \"Retrying connection to CAZy in 10s\"\n",
    "                    )\n",
    "                success = False\n",
    "                response = None\n",
    "                err = err_message\n",
    "            if response is not None:  # response was successful\n",
    "                success = True\n",
    "            # if response from webpage was not successful\n",
    "            tries += 1\n",
    "            time.sleep(10)\n",
    "            \n",
    "        if (not success) or (response is None):\n",
    "            print(f\"Failed to connect to CAZy.\\nError: {err}\")\n",
    "            return None, err\n",
    "        else:\n",
    "            return response, None\n",
    "\n",
    "    return wrapper\n",
    "\n",
    "\n",
    "@browser_decorator\n",
    "def get_page(url, **kwargs):\n",
    "    \"\"\"Create browser and use browser to retrieve page for given URL.\n",
    "\n",
    "    :param url: str, url to webpage\n",
    "    :param args: cmd-line args parser\n",
    "    :kwargs max_tries: max number of times connection to CAZy can be attempted\n",
    "\n",
    "    Return browser response object (the page).\n",
    "    \"\"\"\n",
    "    # create browser object\n",
    "    browser = mechanicalsoup.Browser()\n",
    "    # create response object\n",
    "    page = browser.get(url, timeout=45)\n",
    "    page = page.soup\n",
    "\n",
    "    return page\n",
    "\n",
    "\n",
    "def get_all_accessions(bs_element_lst):\n",
    "    \"\"\"Retrieve all accessions listed in a cell from a CAZyme table.\n",
    "\n",
    "    :param bs_element_list: list of BeautifulSoup element from cell in HTML table\n",
    "\n",
    "    Return list of accessions.\"\"\"\n",
    "    accessions = []\n",
    "\n",
    "    for bs_element in bs_element_lst:\n",
    "        try:\n",
    "            if bs_element.name == \"a\":  # Hyperlinked, extract accession and add to primary\n",
    "                accessions.append(bs_element.text)\n",
    "            elif bs_element.strip() != \"\":  # There is text in element\n",
    "                accessions.append(bs_element)\n",
    "        except TypeError:\n",
    "            pass\n",
    "\n",
    "    return accessions\n",
    "\n",
    "\n",
    "def get_cazy_prots(url, col_index, file_path=None):\n",
    "    \"\"\"Get the NCBI protein accessions for proteins in the structure or characterised tables.\n",
    "    \n",
    "    :param url: str, URL to the CAZy family structure tabel\n",
    "    :param col_index: int, 3 = structure, 4 = characterised\n",
    "    \n",
    "    Return list of NCBI protein accessions\n",
    "    \"\"\"\n",
    "    if file_path is None:\n",
    "        page, error_mss = get_page(\n",
    "            url,\n",
    "            max_tries=100\n",
    "        )\n",
    "        if page is None:\n",
    "            print('Did not retrieve page')\n",
    "            print(error_mss)\n",
    "            return\n",
    "    else:\n",
    "        with open(file_path, \"r\") as fp:\n",
    "            page = BeautifulSoup(fp, features=\"lxml\")\n",
    "    \n",
    "    cazyme_table = page.select('table')[1]\n",
    "\n",
    "    gbk_bs_elements = []\n",
    "\n",
    "    for row in cazyme_table.select(\"tr\"):\n",
    "        try:\n",
    "            if (row.attrs[\"class\"] == ['royaume']) and (row.text.strip() != 'Top'):\n",
    "                continue\n",
    "        except KeyError:\n",
    "            pass\n",
    "\n",
    "        try:\n",
    "            if (row.attrs[\"id\"] == 'line_titre'):\n",
    "                continue\n",
    "        except KeyError:\n",
    "            pass\n",
    "\n",
    "        try:\n",
    "            gbk_bs_elements += [_ for _ in row.select(\"td\")[col_index].contents if getattr(_, \"name\", None) != \"br\"]\n",
    "        except IndexError:\n",
    "            pass\n",
    "\n",
    "    ncbi_accessions = get_all_accessions(gbk_bs_elements)\n",
    "    \n",
    "    ncbi_accessions = list(set(ncbi_accessions))\n",
    "    \n",
    "    return ncbi_accessions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6a6a9df",
   "metadata": {},
   "source": [
    "## Get 'characterised' proteins from CAZy\n",
    "\n",
    "To retrieve the protein listed as 'characterised' in CAZy, for each CAZy family of interest call `get_cazy_prots` and provide:\n",
    "* The URL to the CAZy family characterised page, e.g. 'http://www.cazy.org/PL1_characterized.html'\n",
    "* Provide the index of the html table column containing the NCBI protein accessions, this is index 4 for the characterised table\n",
    "\n",
    "Optionally, if you have saved the HTML page from CAZy, you can provide the path to the HTML file and data can be retrieved directly from this file, e.g. file_path=\"pages/CAZy - PL1.html\". This may be necessary when retrieving data from many CAZy pages, because CAZy will close the connection if too many frequent calls to CAZy are made.\n",
    "\n",
    "`get_cazy_prots()` returns a list of NCBI protein version accessions. \n",
    "\n",
    "Store this list in the `functioned_prots` dict, keyed by the name of the CAZy family and valued by the list of NCBI protein version accessions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ba6c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "functioned_prots = {}\n",
    "functioned_prots['PL1'] = get_cazy_prots('http://www.cazy.org/PL1_characterized.html', 4, file_path=\"pages/CAZy - PL1.html\")\n",
    "functioned_prots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18b42c1c",
   "metadata": {},
   "source": [
    "## Get structurally characterised proteins\n",
    "\n",
    "As with the 'characterised' proteins, proteins in the CAZy structure table for each CAZy family can be retrieved directly from CAZy. See the 'Get 'characterised' proteins from CAZy' above, but use column index 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd0be84",
   "metadata": {},
   "outputs": [],
   "source": [
    "structured_prots = {}\n",
    "structured_prots['PL1'] = get_cazy_prots('http://www.cazy.org/PL1_structure.html', 3)\n",
    "structured_prots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97a5f76b",
   "metadata": {},
   "source": [
    "Not all structurally characterised members of a CAZy family are listed in CAZy. The structurally characterised members of a CAZy family can also be retrieved from UniProt using `cazy_webscraper`.\n",
    "\n",
    "Specifically, use `cw_get_uniprot_data` to retrieve the RCSB PDB ids for proteins in your CAZy families of interest:\n",
    "```\n",
    "cw_get_uniprot_data <path to local cazyme db> --pdb\n",
    "```\n",
    "Use `cw_query_database` to generate a `JSON` file keyed by the NCBI protein version acession and valued by a dict. This `JSON` file can be parsed and added to the `structured_prots` dict.\n",
    "```\n",
    "cw_query_database <path to local cazyme db>\n",
    "```\n",
    "\n",
    "We recommend using data retrieved from CAZy **and** UniProt because there are proteins listed as structurally characterised in CAZy that are not labelled as such in UniProt, and vice versa. Therefore, using data from both databases will provide the most comprehensive representation of the structural characterisation of the family."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7418e977",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_uniprot_structure_prots():\n",
    "    \"\"\"Identify proteins with a least one PDB protein structure ID in UniProt\n",
    "    \n",
    "    :param:\n",
    "    \n",
    "    Return list of NCBI protein version accessions\n",
    "    \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d9d527",
   "metadata": {},
   "outputs": [],
   "source": [
    "structured_prots['PL1'].append(get_uniprot_structure_prots())\n",
    "structured_prots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2955599",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "Load the BLASTP/DIAMOND data into a `pandas` dataframe.\n",
    "\n",
    "CAZy can contain groups of identical proteins. Optionally, these groups can be removed and one representative from group retained, to remove redundant sequences from downstream analyses. When removing redundant proteins, all proteins labelled as candidates, structurally characterised and/or functionally characterised are retained (even if they are redundant).\n",
    "\n",
    "We adivse running the visualisation of the sequence analysis with and without redundant sequences removed.\n",
    "\n",
    "To remove redundant sequences from the dataset, set the `remove_redundant` parameter to `True` when calling `load_data()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc0045f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this block before loading data\n",
    "\n",
    "def load_data(data_file, fam):\n",
    "    \"\"\"Load the output data from BLASTP+/diamond into a pandas dataframe, and calculate the BLAST score ratio.\n",
    "    \n",
    "    :param data_file: str, path to the DIAMOND output file\n",
    "    :param fam: str, name of CAZy family\n",
    "    \n",
    "    Return a pandas dataframe\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(data_file, sep='\\t', header=None)\n",
    "    \n",
    "    column_names = ['qseqid', 'sseqid', 'qlen', 'slen', 'length', 'pident', 'evalue', 'bitscore']\n",
    "    df.columns = column_names\n",
    "    \n",
    "    df['BSR'] = df['bitscore'] / df['qlen']\n",
    "    \n",
    "    df['qcov'] = df['length'] / df['qlen']\n",
    "    df['scov'] = df['length'] / df['slen']\n",
    "    \n",
    "    df = remove_redunant_prots(df, fam)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def remove_redunant_prots(df, fam):\n",
    "    \"\"\"Identify groups of identical proteins, and retaining only one member per group\n",
    "    \n",
    "    :param df: pandas df containing data from DIAMOND\n",
    "    :param fam: str, name of CAZy family - group name in the CANDIDATES df\n",
    "    \n",
    "    return df\n",
    "    \"\"\"\n",
    "\n",
    "    # redundant proteins have a qcov of 1 and a pi of 100\n",
    "    redundance_df = df.loc[ ( (df['pident'] == 100) & (df['qcov'] == 1) ) ]\n",
    "    # find  the groups of redundant proteins\n",
    "    redundant_grps = {}\n",
    "\n",
    "    grp_num = 0\n",
    "\n",
    "    parsed_prots = set()\n",
    "\n",
    "    # identify groups of redundant proteins\n",
    "    for ri in tqdm(range(len(redundance_df)), desc=\"Identifying IPGs\"):\n",
    "        row = redundance_df.iloc[ri]\n",
    "\n",
    "        qseqid = row['qseqid']\n",
    "\n",
    "        if qseqid in parsed_prots:\n",
    "            continue  # has already been added\n",
    "\n",
    "        # get all rows with the same query seq id\n",
    "        qseqid_rows = redundance_df.loc[redundance_df['qseqid'] == qseqid]\n",
    "\n",
    "        if len(qseqid_rows) == 1:\n",
    "            continue  # aligned against self only\n",
    "\n",
    "        subject_ids_to_add = set()\n",
    "\n",
    "        # for each subject id\n",
    "        # check if the versus is true, the qseqid is the sseqid when the sseqid is the qseqid\n",
    "        for q_ri in range(len(qseqid_rows)):\n",
    "            q_row = qseqid_rows.iloc[q_ri]\n",
    "            sub_seqid = q_row['sseqid']\n",
    "\n",
    "            # retrieve the row where the qseqid is now the subject, and the subject id is now the query seq\n",
    "            # They are already in the redundancy df, therefore pident is 100 and qcov is 1\n",
    "            sseqid_rows = redundance_df.loc[(\n",
    "                (redundance_df['qseqid'] ==  sub_seqid) &\n",
    "                (redundance_df['sseqid'] ==  qseqid))\n",
    "            ]\n",
    "\n",
    "            if len(sseqid_rows) > 0:\n",
    "                subject_ids_to_add.add(sub_seqid)\n",
    "\n",
    "        if len(subject_ids_to_add) > 0:\n",
    "            # found redunant pairs for qseqid\n",
    "            redundant_grps[grp_num] = {qseqid}\n",
    "\n",
    "            for sub_seqid in subject_ids_to_add:\n",
    "                redundant_grps[grp_num].add(sub_seqid)\n",
    "                parsed_prots.add(sub_seqid)\n",
    "\n",
    "            grp_num += 1\n",
    "\n",
    "        parsed_prots.add(qseqid)\n",
    "\n",
    "    # from each group select a representative protein\n",
    "    # and identify members of the group that will be dropped\n",
    "    removing = set()\n",
    "    \n",
    "    print(f\"Identified {len(list(redundant_grps.keys()))} groups of identical proteins\")\n",
    "\n",
    "    for grp in redundant_grps:\n",
    "        prots_to_keep = set{}\n",
    "        \n",
    "        for prot in redundant_grps[grp]:\n",
    "            try:\n",
    "                # retain proteins marked as candidates, functionally characitersed or structurally characterised\n",
    "                if prot in CANDIDATES[fam]:\n",
    "                    prots_to_keep.add(prot)\n",
    "                elif prot in structured_prots[fam]:\n",
    "                    prots_to_keep.add(prot)\n",
    "                elif prot in functioned_prots[fam]:\n",
    "                    prots_to_keep.add(prot)\n",
    "                elif len(prots_to_keep) == 0: # ensure at least one protein from the group is retained\n",
    "                    prots_to_keeps.add(prot)\n",
    "                else:  # already have members from the group so drop the protein\n",
    "                    removing.add(prot)\n",
    "            except KeyError:\n",
    "                if len(prots_to_keep) == 0:\n",
    "                    prots_to_keeps.add(prot)\n",
    "                else:  # already have members from the group so drop the protein\n",
    "                    removing.add(prot)\n",
    "\n",
    "    df = df[~df['qseqid'].isin(removing)]\n",
    "    df = df[~df['sseqid'].isin(removing)]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d76a9e74",
   "metadata": {},
   "source": [
    "## Visualise sequence diversity\n",
    "\n",
    "This notebook generates a clustermap based upon a user defined variable. Heatmaps can also be generated to plot alternative variables, with proteins listed in the same order on the x and y-axis as the clustermap, to facilitate comparison of multiple variables.\n",
    "\n",
    "We recommend:\n",
    "* Generating a clustermap using the BLAST Score Ratio (BSR)\n",
    "* Generating a heatmap of the percentage identiy with proteins listed in the same order as the clustermap\n",
    "* Generating a heatmap of the query covereage with proteins listed in the same order as the clustermap\n",
    "\n",
    "**Annotating candidates and functionally/structurally characterised proteins:**  \n",
    "To generate a clustermap without annotating functionall/structurally characterised proteins, do not use the `annotate` parameter when calling `plot_clustermap()` and `plot_heatmap_of_clustermap()`. To annotate functionally/structurally characterised proteins set the `annotate` parameter to `True`.\n",
    "\n",
    "**Plotting only candidates and functionally/structurally characterised proteins:**\n",
    "To plot all proteins in a group (i.e. a CAZy family) do not call the `char_only` parameter when calling `plot_clustermap()` and `plot_heatmap_of_clustermap()`. To plot only the candidates and functionally/structurally characterised proteins, set the `char_only` parameter to `True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53d3c94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this block before visualising the data\n",
    "\n",
    "def plot_clustermap(\n",
    "    df,\n",
    "    fam,\n",
    "    varaible,\n",
    "    title=None,\n",
    "    colour_scheme='rocket_r',\n",
    "    fig_size=(25, 25),\n",
    "    save_fig=None,\n",
    "    dpi=100,\n",
    "    annotate=False,\n",
    "    char_only=False,\n",
    "):\n",
    "    \"\"\"Plot a cluster map for the specified variable\n",
    "    \n",
    "    :param df: pandas dataframe\n",
    "    :param fam: str, CAZy family of interest\n",
    "    :param variable: df, name of column containing the variable to plot\n",
    "    :param title: str, default none. Title of plot\n",
    "    :param colour_scheme: str, default rocket_r, seaborn colour scheme for plot\n",
    "    :param fig_size: tuple, len 2, default (25, 10)\n",
    "    :param save_fig: str, path to save file, default none, don't save fig\n",
    "    :param dpi: int, default 100, resolution of saved file image\n",
    "    :param annotate: bool, add annotation of protein candidates, and functionally/structurally \n",
    "        characteirsed proteins\n",
    "    :param char_only: bool, if set to true, only plot proteins labelled as candidates or\n",
    "        functionally/structurally characteirsed proteins\n",
    "    \n",
    "    Return seaborn plot\n",
    "    \"\"\"\n",
    "    df = df[['qseqid', 'sseqid', varaible]]\n",
    "    \n",
    "    if char_only:  # plot only proteins that are candidates and functionally/structurally characteirsed proteins\n",
    "        charactised_prots = functioned_prots[fam] + structured_prots[fam] + CANDIDATES[fam]\n",
    "        df = df[df['qseqid'].isin(charactised_prots)]\n",
    "        df = df[df['sseqid'].isin(charactised_prots)]\n",
    "    \n",
    "    heatmap_data = pd.pivot_table(df, index='qseqid', columns='sseqid', values=varaible)\n",
    "    heatmap_data.columns = list(heatmap_data.columns)\n",
    "    heatmap_data.index = list(heatmap_data.columns)\n",
    "    heatmap_data = heatmap_data.fillna(0)\n",
    "    \n",
    "    if annotate:\n",
    "        # add extra info on structural and functional characterisation of the family\n",
    "        extra_data = []\n",
    "\n",
    "        for prot in list(heatmap_data.columns):\n",
    "            if prot in CANDIDATES[fam]:\n",
    "                if prot in functioned_prots[fam]:\n",
    "                    extra_data.append(STRUCT_PAL_DICT['funcCand'])\n",
    "                elif prot in structured_prots[fam]:\n",
    "                    extra_data.append(STRUCT_PAL_DICT['structCand'])\n",
    "                else:\n",
    "                    extra_data.append(STRUCT_PAL_DICT['cand'])\n",
    "\n",
    "            elif prot in structured_prots[fam]:\n",
    "                extra_data.append(STRUCT_PAL_DICT['struct'])\n",
    "\n",
    "            elif prot in functioned_prots[fam]:\n",
    "                extra_data.append(STRUCT_PAL_DICT['func'])\n",
    "\n",
    "            else:\n",
    "                extra_data.append(STRUCT_PAL_DICT['nothing'])\n",
    "\n",
    "        fig = sns.clustermap(\n",
    "            heatmap_data,\n",
    "            cmap=colour_scheme,\n",
    "            figsize=fig_size,\n",
    "            row_colors=extra_data,\n",
    "            col_colors=extra_data,\n",
    "        );\n",
    "\n",
    "        # extra data legend\n",
    "        for label in list(STRUCT_PAL_DICT.keys()):\n",
    "            fig.ax_row_dendrogram.bar(0, 0, color=STRUCT_PAL_DICT[label], label=label, linewidth=0)\n",
    "\n",
    "        l3 = fig.ax_row_dendrogram.legend(title='Characterisation', loc='upper right', ncol=1)\n",
    "    \n",
    "    else:\n",
    "        fig = sns.clustermap(\n",
    "            heatmap_data,\n",
    "            cmap=colour_scheme,\n",
    "            figsize=fig_size,\n",
    "        );\n",
    "    \n",
    "    if save_fig is not None:\n",
    "        fig.savefig(save_fig, dpi=dpi);\n",
    "    \n",
    "    return fig\n",
    "\n",
    "\n",
    "def plot_heatmap_of_clustermap(\n",
    "    fig,\n",
    "    df,\n",
    "    fam,\n",
    "    varaible,\n",
    "    title=None,\n",
    "    colour_scheme='rocket_r',\n",
    "    fig_size=(25, 25),\n",
    "    save_fig=None,\n",
    "    dpi=100,\n",
    "    annotate=False,\n",
    "    char_only=False,\n",
    "):\n",
    "    \"\"\"Generate a heatmap for the defined variable, with proteins plotted in the same order as the provided\n",
    "    clustermap (fig)\n",
    "    \n",
    "    :param fig: seaborn clustergrid of entire family, default None, clustermap,\n",
    "    :param df: pandas dataframe\n",
    "    :param fam: str, CAZy family of interest\n",
    "    :param variable: df, name of column containing the variable to plot\n",
    "    :param title: str, default none. Title of plot\n",
    "    :param colour_scheme: str, default rocket_r, seaborn colour scheme for plot\n",
    "    :param fig_size: tuple, len 2, default (25, 10)\n",
    "    :param save_fig: str, path to save file, default none, don't save fig\n",
    "    :param dpi: int, default 100, resolution of saved file image\n",
    "    :param annotate: bool, add annotation of protein candidates, and functionally/structurally \n",
    "        characteirsed proteins\n",
    "    :param char_only: bool, if set to true, only plot proteins labelled as candidates or\n",
    "        functionally/structurally characteirsed proteins\n",
    "    \n",
    "    Return nothing\n",
    "    \"\"\"\n",
    "    column_order = list(fig.__dict__['data2d'].keys())\n",
    "    row_order = list(fig.__dict__['data2d'].index)\n",
    "    \n",
    "    df = df[['qseqid', 'sseqid', varaible]]\n",
    "    \n",
    "    if char_only:  # plot only proteins that are candidates and functionally/structurally characteirsed proteins\n",
    "        charactised_prots = functioned_prots[fam] + structured_prots[fam] + CANDIDATES[fam]\n",
    "        df = df[df['qseqid'].isin(charactised_prots)]\n",
    "        df = df[df['sseqid'].isin(charactised_prots)]\n",
    "    \n",
    "    heatmap_data = pd.pivot_table(df, index='qseqid', columns='sseqid', values=varaible)\n",
    "    heatmap_data.columns = list(heatmap_data.columns)\n",
    "    heatmap_data.index = list(heatmap_data.columns)\n",
    "    heatmap_data = heatmap_data.fillna(0)\n",
    "    \n",
    "    heatmap_data = heatmap_data.to_dict()  # {col: {row: value}}\n",
    "\n",
    "    heatmap_df_data = {}\n",
    "\n",
    "    for _prot in column_order:\n",
    "        column_data = heatmap_data[_prot] # dict of {row: value} for the column\n",
    "        \n",
    "        for __prot in row_order:\n",
    "            row_value = column_data[__prot]\n",
    "\n",
    "            try:\n",
    "                heatmap_df_data[_prot]  # column\n",
    "            except KeyError:\n",
    "                heatmap_df_data[_prot] = {}\n",
    "\n",
    "            heatmap_df_data[_prot][__prot] = row_value\n",
    "            \n",
    "    if annotate:\n",
    "        # add extra info on structural and functional characterisation of the family\n",
    "        extra_data_col = []\n",
    "\n",
    "        for prot in column_order:\n",
    "            # candidate 1, funct candidate 0.75, structured 0.5, functional 0.25, nothing 0\n",
    "            if prot in CANDIDATES[fam]:\n",
    "                if prot in functioned_prots[fam]:\n",
    "                    extra_data_col.append(STRUCT_PAL_DICT['funcCand'])\n",
    "                elif prot in structured_prots[fam]:\n",
    "                    extra_data_col.append(STRUCT_PAL_DICT['structCand'])\n",
    "                else:\n",
    "                    extra_data_col.append(STRUCT_PAL_DICT['cand'])\n",
    "\n",
    "            elif prot in structured_prots[fam]:\n",
    "                extra_data_col.append(STRUCT_PAL_DICT['struct'])\n",
    "\n",
    "            elif prot in functioned_prots[fam]:\n",
    "                extra_data_col.append(STRUCT_PAL_DICT['func'])\n",
    "\n",
    "            else:\n",
    "                extra_data_col.append(STRUCT_PAL_DICT['nothing'])\n",
    "\n",
    "        extra_data_row = []\n",
    "\n",
    "        for prot in row_order:\n",
    "            # candidate 1, funct candidate 0.75, structured 0.5, functional 0.25, nothing 0\n",
    "            if prot in CANDIDATES[fam]:\n",
    "                if prot in functioned_prots[fam]:\n",
    "                    extra_data_row.append(STRUCT_PAL_DICT['funcCand'])\n",
    "                elif prot in structured_prots[fam]:\n",
    "                    extra_data_row.append(STRUCT_PAL_DICT['structCand'])\n",
    "                else:\n",
    "                    extra_data_row.append(STRUCT_PAL_DICT['cand'])\n",
    "\n",
    "            elif prot in structured_prots[fam]:\n",
    "                extra_data_row.append(STRUCT_PAL_DICT['struct'])\n",
    "\n",
    "            elif prot in functioned_prots[fam]:\n",
    "                extra_data_row.append(STRUCT_PAL_DICT['func'])\n",
    "\n",
    "            else:\n",
    "                extra_data_row.append(STRUCT_PAL_DICT['nothing'])\n",
    "\n",
    "        fig = sns.clustermap(\n",
    "            heatmap_df_data,\n",
    "            cmap=colour_scheme,\n",
    "            figsize=fig_size,\n",
    "            row_cluster=False,\n",
    "            col_cluster=False,\n",
    "            row_colors=extra_data_row,\n",
    "            col_colors=extra_data_col,\n",
    "        );            \n",
    "\n",
    "        # extra data legend\n",
    "        for label in list(STRUCT_PAL_DICT.keys()):\n",
    "            fig.ax_row_dendrogram.bar(0, 0, color=STRUCT_PAL_DICT[label], label=label, linewidth=0)\n",
    "\n",
    "        l3 = fig.ax_row_dendrogram.legend(title='Info', loc='upper right', ncol=1)\n",
    "    \n",
    "    else:\n",
    "        fig = sns.clustermap(\n",
    "            heatmap_df_data,\n",
    "            cmap=colour_scheme,\n",
    "            figsize=fig_size,\n",
    "            row_cluster=False,\n",
    "            col_cluster=False,\n",
    "        );\n",
    "\n",
    "    if save_fig is not None:\n",
    "        fig.savefig(save_fig, dpi=dpi);\n",
    "    \n",
    "    fig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d410faf",
   "metadata": {},
   "source": [
    "## Family analysis\n",
    "\n",
    "Here is some example code for running the analysis for CAZy family PL20.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70626b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pl20_df = load_diamond_data('pl20_blastp_out', 'PL20')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce71225d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pl20_bsr_plt = plot_clustermap(pl20_df, 'PL20', 'BSR', fig_size=(100, 100), save_fig='pl20_clustermap.png')\n",
    "pl20_bsr_plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cd53731",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot a clustermap of only the candidates and functionally/structurally characterised proteins\n",
    "# that is also annotated to differentiate, candidates and functionally/structurally characterised proteins\n",
    "pl20_char_bsr_plt = plot_clustermap(\n",
    "    pl4_df,\n",
    "    'PL20',\n",
    "    'BSR',\n",
    "    fig_size=(7, 7),\n",
    "    save_fig='pl20_clustermap_char.png',\n",
    "    char_only=True,\n",
    "    annotate=True,\n",
    ")\n",
    "pl20_char_bsr_plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57e3e920",
   "metadata": {},
   "source": [
    "Then plot the precentage identity and query coverage for the candidate and functionally/structurally characterised proteins, plotting the proteins on the heatmaps in the same order as they appear in the clustermap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1571405e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('PL20 percentage identity, colour scheme blue')\n",
    "plot_heatmap_of_clustermap(\n",
    "    pl20_char_bsr_plt,\n",
    "    pl20_df,\n",
    "    'PL20',\n",
    "    'pident',\n",
    "    fig_size=(7, 7),\n",
    "    save_fig='pl20_clustermap_PIDENT_char.png',\n",
    "    colour_scheme=sns.color_palette(\"Blues\", as_cmap=True),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e01fb64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('PL20 query coverage, colour scheme purple')\n",
    "plot_heatmap_of_clustermap(\n",
    "    pl20_char_bsr_plt,\n",
    "    pl20_df,\n",
    "    'PL20',\n",
    "    'qcov',\n",
    "    fig_size=(7, 7),\n",
    "    save_fig='pl20_clustermap_QCOV_char.png',\n",
    "    colour_scheme=sns.color_palette(\"Purples\", as_cmap=True),\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
